### PR TITLE
AI consider enemy turn order

### PR DESCRIPTION
**Pull request by ron-murhammer**
[Review on Reviewable](https://reviewable.io/reviews/triplea-game/triplea/282)

### CODE REVIEW

File: src/games/strategy/triplea/ai/proAI/util/ProAttackOptionsUtils.java [hunk 1]
```diff
@@ -396,6 +396,18 @@ public void findAttackOptions(final PlayerID player, final List<Territory> myUni
       final Map<Territory, Set<Territory>> landRoutesMap, final List<ProAmphibData> transportMapList,
       final List<Territory> enemyTerritories, final List<Territory> territoriesToCheck,
       final boolean isCheckingEnemyAttacks, final boolean isIgnoringRelationships) {
+    findAttackOptions(player, myUnitTerritories, moveMap, unitMoveMap, transportMoveMap, bombardMap, landRoutesMap,
+        transportMapList, enemyTerritories, new ArrayList<Territory>(), territoriesToCheck, isCheckingEnemyAttacks,
+        isIgnoringRelationships);
+  }
+
+  public void findAttackOptions(final PlayerID player, final List<Territory> myUnitTerritories,
```
Comment (DanVanAtta) [inline src/games/strategy/triplea/ai/proAI/util/ProAttackOptionsUtils.java]: Keep this as one method rather than splitting it up as two? I'm also unsure why the method would be part of the public API and not private, but re-merging it back to one would make that a moot point.

Comment (ron-murhammer) [inline src/games/strategy/triplea/ai/proAI/util/ProAttackOptionsUtils.java]: Agree. I ended up already doing this in the next set of changes. Really just did this to avoid having to update any references that called the original method without the extra parameter. I'd prefer to just merge this one as I have lots of refactoring changes already made in my next PR.

File: src/games/strategy/triplea/ai/proAI/util/ProAttackOptionsUtils.java [hunk 2]
```diff
@@ -445,23 +458,29 @@ public void findDefendOptions(final PlayerID player, final List<Territory> myUni
...
     findAirMoveOptions(player, myUnitTerritories, moveMap, unitMoveMap,
-        ProMatches.territoryIsNotConqueredAlliedLand(player, data), new ArrayList<Territory>(), false, false, false);
+        ProMatches.territoryIsNotConqueredAlliedLand(player, data), new ArrayList<Territory>(),
+        new ArrayList<Territory>(), false, false, false);
     findAmphibMoveOptions(player, myUnitTerritories, moveMap, transportMapList, landRoutesMap,
         Matches.isTerritoryAllied(player, data), new ArrayList<Territory>(), false, false, false);
   }
 
-  public void findMaxEnemyAttackUnits(final PlayerID player, final List<Territory> myConqueredTerritories,
+  public void findEnemyAttackOptions(final PlayerID player, final List<Territory> myConqueredTerritories,
```
Comment (DanVanAtta) [inline src/games/strategy/triplea/ai/proAI/util/ProAttackOptionsUtils.java]: Check to see if you can use a Set instead of List as parameter types. You convert a lot of sets into ArrayLists when calling this function, looks like that is more common than passing in an actual list.

Comment (DanVanAtta) [inline src/games/strategy/triplea/ai/proAI/util/ProAttackOptionsUtils.java]: Would be nice to see this method have a return type rather than rely on out parameters, side effects. Secondly, making it static if possible would also help, at that point it would not be stateful anymore and generally non-stateful methods are simpler. Could also write a test for it more easily as well.

File: src/games/strategy/triplea/ai/proAI/util/ProUtils.java [hunk 3]
```diff
@@ -46,6 +47,26 @@ public ProUtils(final ProAI ai) {
     return unitTerritoryMap;
   }
 
+  public List<PlayerID> getEnemyPlayersInTurnOrder(final PlayerID player) {
+    final GameData data = ai.getGameData();
+    final List<PlayerID> enemyPlayers = new ArrayList<PlayerID>();
+    GameSequence sequence = data.getSequence();
```
Comment (DanVanAtta) [inline src/games/strategy/triplea/ai/proAI/util/ProUtils.java]: inconsistent use of optional `final`. Since it's used in the above two lines, it looks intentional, but at the same time it looks like the variables could final, so the inconsistency causes some small confusion.

File: src/games/strategy/triplea/ai/proAI/util/ProUtils.java [hunk 4]
```diff
@@ -46,6 +47,26 @@ public ProUtils(final ProAI ai) {
     return unitTerritoryMap;
   }
 
+  public List<PlayerID> getEnemyPlayersInTurnOrder(final PlayerID player) {
```
Comment (DanVanAtta) [inline src/games/strategy/triplea/ai/proAI/util/ProUtils.java]: Very testable method, a test case would help you make sure you don't have any regressions in case the logic needs to change for whatever reason. Please consider adding  a test case for this new code.

File: src/games/strategy/triplea/ai/proAI/util/ProAttackOptionsUtils.java (hunk 2 above)
Comment (ron-murhammer) [inline src/games/strategy/triplea/ai/proAI/util/ProAttackOptionsUtils.java]: Agree with both points. In general, I need to use Sets instead of Lists more often. I just tend to overuse Lists because I like their API better and also didn't realize a bunch of them should be Sets. There are also lots of existing 'engine methods' that use Lists so it requires converting somewhere.

I would like to make all the utils classes static. I was going a different direction originally and going to save state in them but most became just pure utility classes. It requires a bunch of refactoring though due to how GameData is shared right now.

**Comment by DanVanAtta** [comment]
- There are some significant code quality issues that IMO need to be addressed before much else can be done. Essentially a few rounds of refactoring, adding unit/tregression tests, intermediary data structures and breaking up long methods.

- The one "key" new method added needs a unit test to make sure you covered your "Ps" and "Qs" and so we have a regression test in place

- I can't tell if strength of enemy players is considered or not when considering if an enemy can move over a newly captured territory. I also can't tell if allied counter attacks/counter captures are handled correctly. Essentially the question would be, if you had a stack of german units one space away from moscow, and you pull up one italian tank, would the Russian AI overreact thinking that it can't stop the can-opener? The answer to that question changes if UK is around for example to re-capture the territory. 

- If last point is addressed, then sweet. Otherwise beyond the code quality issues this does look like a step in the right direction in terms of functionality.

**Comment by DanVanAtta** [comment]
> " I'd prefer to just merge this one as I have lots of refactoring changes already made in my next PR."

If it helps and will make the review easier, then please go for it. Otherwise personally I would like to keep this open for a bit as a reminder to add the unit test for the new code.

File: src/games/strategy/triplea/ai/proAI/util/ProUtils.java (hunk 3 above)
Comment (ron-murhammer) [inline src/games/strategy/triplea/ai/proAI/util/ProUtils.java]: Good catch. I usually just leave on the auto-format setting to add final where it can be added. Looks like that setting was turned off during the changing of the formatting profile.

File: src/games/strategy/triplea/ai/proAI/util/ProUtils.java (hunk 4 above)
Comment (ron-murhammer) [inline src/games/strategy/triplea/ai/proAI/util/ProUtils.java]: I would say "almost" very testable method. The issue is this "final GameData data = ai.getGameData();". So you need to setup the full AI with initialized game data to test this. This is the big issue right now as I have essentially everything extending from ProAI main class which holds the current game data. This is beneficial in that it doesn't need passed around as parameters but the negative is that they can't be static and are all tied back to that class.

I also ended up completed refactoring this method as well since I needed to create methods to get allied players and all players in turn order.

**Comment by ron-murhammer** [comment]
1&2. So both refactoring and unit tests are really lacking right now in the AI code. The biggest reason for this is the scope of what really needs done is pretty large and I really haven't had much time recently to code. I know its not a great excuse but I'd like to refactor significantly before writing unit tests for any AI code mostly just to save time (and yes I know one of the points of unit tests is to be able to refactor with lower risk of breaking functionality).

3. So for this initial implementation, the AI is overly conservative with can-opener attacks. It essentially considers that even just one unit can clear any territory as you describe above and it doesn't rely on any of its allies to block since these then rely on what the enemies do (you get into starting to have to simulate out turns which given the complexity isn't possible with today's compute power). Its somewhat more complicated then you describe since the current enemy attack options are calculated before the AI makes its non-combat moves so it doesn't know how many units it has to block potential can-openers yet. It essentially would need to recalculate enemy attack options after it moves each of its units in order to consider if a territory can be blocked or not. This again would take a significant amount of processing power for probably somewhat minimal gains.

To summarize, the AI plays conservative around can-openers to avoid losing its key factories and capital with this initial implementation and there are definitely ways to improve it but processing time needs to be considered.

4. So I think I'm going to merge this since I tested it and think the cod quality is probably about equal  to the existing (so better functionality and equally poor code quality in the AI). I have a few more commits that I'll be making a PR for soon that add more turn order functionality but start tackling some of the refactoring.

File: src/games/strategy/triplea/ai/proAI/util/ProUtils.java (hunk 4 above)
Comment (DanVanAtta) [inline src/games/strategy/triplea/ai/proAI/util/ProUtils.java]: Pretty easy to mock the AI and gameData with mockito.
```
@Mock
GameData mockGameData
@Mock
ProAI mockAI

when( mockGameData.getGameData()).thenReturn(mockAI)
when(mockAI...()).thenReturn()
```

Since the interactions are limited to just the `ai`, the mocking is pretty straight forward. The single AI object is not a bad thing IMO, it hides the data and provides some utility around it. Plus it actually can make for decent mocking since you can use some pretty decent high level functionality with it, better than just having primitives to work with.

**Comment by DanVanAtta** [comment]
unit tests are effectively regression tests. Personally I would take out abstract out chunks of code wholesale into new modules that have nice API's, test those, then reduce the code and refactor more. Repeating enough times and the original code gets to be higher level and dependent on only stuff that has been tested.

I've learned the hard way that you go faster when you clean the code first then do stuff. Working with dirty code and muddling through is just slower in the long run, generally hitting the few bugs and getting really slowed down by them can be the difference, and the worst case is more extreme. 

Once code is written though IMO you lose quite a bit of value from tests. They are good for telling you when you are done. For example in above, a stub interface and then writing tests that fail against an empty implementation would be a good way to generate tests before the implementation was copied over.

*note: sorry if this is rough, didn't go through and edit it in favor of time.

File: src/games/strategy/triplea/ai/proAI/util/ProUtils.java (hunk 4 above)
Comment (DanVanAtta) [inline src/games/strategy/triplea/ai/proAI/util/ProUtils.java]: This class has examples of mockito: https://github.com/triplea-game/triplea/blob/a1554a51f4033105dd500759a14e72c6783263fd/test/games/strategy/engine/framework/startup/mc/GameSelectorModelTest.java

Comment (ron-murhammer) [inline src/games/strategy/triplea/ai/proAI/util/ProUtils.java]: True. Though there are other problems as well. Having all the utilities as objects is causing a giant web of parameters to pass around. I think what I want to do is create a sort of static class to hold the game data and static parameters around it such as settings, players, etc (almost how the AI would translate the game.xml and current game data).

Comment (DanVanAtta) [inline src/games/strategy/triplea/ai/proAI/util/ProUtils.java]: IMO you're describing the problem that dependency injection is meant to be solve. I think it's time we start incorporating Dagger (we had discussed IOC containers earlier, Dagger was a stronger candidate than others, such as Spring for example)
